File: Failang/compiler/parser/parser.py
from compiler.lexer.tokens import TokenType
from compiler.ast.nodes import Program, VariableDeclaration, PrintStatement, WhileStatement, BinaryOperation, Number, Variable

class Parser:
    def __init__(self, tokens): self.tokens = tokens; self.position = 0
    def current(self): return self.tokens[self.position] if self.position < len(self.tokens) else None
    def advance(self): self.position += 1
    def parse(self):
        stmts = []
        while self.current() and self.current().type != TokenType.EOF:
            stmt = self.statement()
            if stmt: stmts.append(stmt)
        return Program(stmts)
    def statement(self):
        token = self.current()
        if not token: return None
        if token.type == TokenType.LET:
            self.advance(); name = self.current().value; self.advance(); self.advance(); val = self.expression()
            return VariableDeclaration(name, val)
        if token.type == TokenType.PRINT:
            self.advance(); return PrintStatement(self.expression())
        if token.type == TokenType.WHILE:
            self.advance(); cond = self.expression(); self.advance() # LBRACE
            body = []
            while self.current() and self.current().type != TokenType.RBRACE: body.append(self.statement())
            self.advance() # RBRACE
            return WhileStatement(cond, body)
        self.advance(); return None
    def expression(self):
        left = self.primary()
        if self.current() and self.current().type in [TokenType.PLUS, TokenType.LESS]:
            op = self.current(); self.advance(); return BinaryOperation(left, op, self.primary())
        return left
    def primary(self):
        token = self.current(); self.advance()
        if token.type == TokenType.NUMBER: return Number(token.value)
        if token.type == TokenType.IDENTIFIER: return Variable(token.value)
        return None

File: src/compiler/parser.rs
use crate::compiler::tokens::{Token, TokenKind};
use crate::compiler::ast::{Expr, Stmt, LiteralValue};

pub struct Parser {
    tokens: Vec<Token>,
    current: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, current: 0 }
    }

    pub fn parse(&mut self) -> Vec<Stmt> {
        let mut statements = Vec::new();
        while !self.is_at_end() {
            if let Some(stmt) = self.declaration() {
                statements.push(stmt);
            }
        }
        statements
    }

    fn declaration(&mut self) -> Option<Stmt> {
        if self.match_token(TokenKind::Let) {
            self.var_declaration()
        } else {
            self.statement()
        }
    }

    fn var_declaration(&mut self) -> Option<Stmt> {
        let name = if let TokenKind::Identifier(n) = &self.peek().kind {
            n.clone()
        } else {
            panic!("يتوقع وجود اسم متغير بعد كلمة 'متغير' في السطر {}", self.peek().line);
        };
        
        self.advance();
        self.consume(TokenKind::Assign, "يتوقع وجود '=' بعد اسم المتغير");
        
        let initializer = self.expression();
        Some(Stmt::Var { name, initializer })
    }

    fn statement(&mut self) -> Option<Stmt> {
        if self.match_token(TokenKind::Print) {
            let expr = self.expression();
            Some(Stmt::Print(expr))
        } else {
            let expr = self.expression();
            Some(Stmt::Expression(expr))
        }
    }

    fn expression(&mut self) -> Expr {
        self.equality()
    }

    fn equality(&mut self) -> Expr {
        let mut expr = self.term();
        while self.match_token(TokenKind::Plus) || self.match_token(TokenKind::Minus) {
            let operator = self.previous().clone();
            let right = self.term();
            expr = Expr::Binary {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            };
        }
        expr
    }

    fn term(&mut self) -> Expr {
        let token = self.advance().clone();
        match token.kind {
            TokenKind::Number(n) => Expr::Literal(LiteralValue::Number(n)),
            TokenKind::String(s) => Expr::Literal(LiteralValue::String(s)),
            TokenKind::Identifier(name) => Expr::Variable(name),
            _ => panic!("تعبير غير صالح في السطر {}", token.line),
        }
    }

    fn match_token(&mut self, kind: TokenKind) -> bool {
        if self.check(kind) {
            self.advance();
            return true;
        }
        false
    }

    fn consume(&mut self, kind: TokenKind, message: &str) {
        if self.check(kind) {
            self.advance();
        } else {
            panic!("{}", message);
        }
    }

    fn check(&self, kind: TokenKind) -> bool {
        if self.is_at_end() { return false; }
        std::mem::discriminant(&self.peek().kind) == std::mem::discriminant(&kind)
    }

    fn advance(&mut self) -> &Token {
        if !self.is_at_end() { self.current += 1; }
        self.previous()
    }

    fn is_at_end(&self) -> bool {
        matches!(self.peek().kind, TokenKind::EOF)
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.current]
    }

    fn previous(&self) -> &Token {
        &self.tokens[self.current - 1]
    }
}
